presto-main/src/test/java/com/facebook/presto/operator/scalar/BenchmarkArraysOverlap.java

```diff
@@ -0,0 +1,154 @@
+import com.facebook.presto.spi.relation.CallExpression;
+import com.facebook.presto.spi.relation.RowExpression;
+import com.facebook.presto.sql.gen.ExpressionCompiler;
+import com.facebook.presto.sql.gen.PageFunctionCompiler;
+import com.google.common.collect.ImmutableList;
+import org.openjdk.jmh.annotations.Benchmark;
+import org.openjdk.jmh.annotations.BenchmarkMode;
+import org.openjdk.jmh.annotations.Fork;
+import org.openjdk.jmh.annotations.Mode;
+import org.openjdk.jmh.annotations.OperationsPerInvocation;
+import org.openjdk.jmh.annotations.OutputTimeUnit;
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.runner.Runner;
+import org.openjdk.jmh.runner.options.Options;
+import org.openjdk.jmh.runner.options.OptionsBuilder;
+import org.openjdk.jmh.runner.options.VerboseMode;
+import org.testng.annotations.Test;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.concurrent.TimeUnit;
+
+import static com.facebook.presto.common.type.BigintType.BIGINT;
+import static com.facebook.presto.memory.context.AggregatedMemoryContext.newSimpleAggregatedMemoryContext;
+import static com.facebook.presto.metadata.MetadataManager.createTestMetadataManager;
+import static com.facebook.presto.sql.analyzer.TypeSignatureProvider.fromTypes;
+import static com.facebook.presto.sql.relational.Expressions.field;
+import static com.facebook.presto.testing.TestingConnectorSession.SESSION;
+
+@SuppressWarnings("MethodMayBeStatic")
+@State(Scope.Thread)
+@OutputTimeUnit(TimeUnit.NANOSECONDS)
+@Fork(3)
+@BenchmarkMode(Mode.AverageTime)
+public class BenchmarkArraysOverlap
+{
+    private static final int POSITIONS = 100_000;
+    private static final int ARRAY_SIZE = 10;
+
+    @Benchmark
+    @OperationsPerInvocation(POSITIONS * ARRAY_SIZE)
+    public List<Optional<Page>> benchmark(BenchmarkData data)
+    {
+        return ImmutableList.copyOf(
+                data.getPageProcessor().process(
+                        SESSION.getSqlFunctionProperties(),
+                        new DriverYieldSignal(),
+                        newSimpleAggregatedMemoryContext().newLocalMemoryContext(PageProcessor.class.getSimpleName()),
+                        data.getPage()));
+    }
+
+    @SuppressWarnings("FieldMayBeFinal")
+    @State(Scope.Thread)
+    public static class BenchmarkData
+    {
+        private Page page;
+        private PageProcessor pageProcessor;
+
+        @Setup
+        public void setup()
+        {
+            MetadataManager metadata = createTestMetadataManager();
+            FunctionAndTypeManager functionAndTypeManager = metadata.getFunctionAndTypeManager();
+            FunctionHandle functionHandle = functionAndTypeManager.lookupFunction("arrays_overlap", fromTypes(new ArrayType(BIGINT), new ArrayType(BIGINT)));
+
+            List<RowExpression> projections = ImmutableList.of(
```

Comment:
Please test all types.

Reply:
What do you feel will be the threshold to switch nlogn? Do you want to add them as constant in the class itself?

Reply:
* Tested all types.
* Also used threshold (1000) to switch to O(N log N) solution based on TypedSet. (Used TypedSet since it uses minimal temp arrays. Also, after we optimized the block builder we may get better performance. Also, makes the code much simple and compact.

Approach (Nested loop + Typed Set - O(N log N))

```
Benchmark                              Mode  Cnt    Score   Error  Units
BenchmarkArraysOverlap.largeBenchmark  avgt   20  101.916 ± 3.431  ns/op
BenchmarkArraysOverlap.smallBenchmark  avgt   20   17.060 ± 1.049  ns/op
```

Approach (Quicksort + temp arrays)

```
Benchmark                              Mode  Cnt    Score   Error  Units
BenchmarkArraysOverlap.largeBenchmark  avgt   20   96.646 ± 3.153  ns/op
BenchmarkArraysOverlap.smallBenchmark  avgt   20  119.372 ± 1.524  ns/op
```

Reply:
The old implementation is faster on large arrays than on smaller arrays? 😮

Reply:
It was interesting to me as well. I was not sure, if that's noise in measurements when I was doing in local.

Reply:
Since a monotonic function with respect to time is expected.

Reply:
> The old implementation is faster on large arrays than on smaller arrays? 😮

it uses qucksort so not a surprise?